File: 0956-tallest-billboard/0956-tallest-billboard.py
class Solution:
    def tallestBillboard(self, rods: List[int]) -> int:
        dp = {0 :0}
        for x in rods:
            d = dp.copy()
            for f,s in d.items():
                dp[f+x] = max(dp.get(x+f,0),s)
                dp[abs(f-x)] = max(dp.get(abs(x-f),0),s+min(f,x))
        return dp.get(0)

File: 0956-tallest-billboard/0956-tallest-billboard.cpp
class Solution {
public:
    int tallestBillboard(vector<int>& rods) {
        unordered_map<int,int>dp;
        dp[0]=0;
        for(int &x:rods){
            unordered_map<int,int>temp=dp;
            for (auto it:temp){
                int d = it.first;
                dp[d+x] = max(dp[d+x],temp[d]);
                dp[abs(d-x)] = max(dp[abs(d-x)],temp[d]+min(d,x));
            }
        }
        return dp[0];
    }
};